### PR TITLE
In-App Updates: Add Version model

### PR DESCRIPTION
**Pull request by momo-ozawa**
Fixes https://github.com/wordpress-mobile/WordPress-iOS/pull/23195#pullrequestreview-2057967801

## Description
- Adds the Version model, which has more robust comparison logic

## How to test
- Make sure `AppUpdateCoordinatorTests` is all green

## Regression Notes
1. Potential unintended areas of impact
n/a

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/Services/AppUpdate/AppUpdateCoordinator.swift [hunk 1]
```diff
@@ -162,12 +172,23 @@ private enum Constants {
...
 
-private extension String {
-    func isLower(than anotherVersionString: String) -> Bool {
-        self.compare(anotherVersionString, options: .numeric) == .orderedAscending
-    }
-
-    func isLowerThanOrEqual(to anotherVersionString: String) -> Bool {
-        [ComparisonResult.orderedSame, .orderedAscending].contains(self.compare(anotherVersionString, options: .numeric))
+ extension Version {
+    init?(from versionString: String) {
```
Comment (kean) [inline WordPress/Classes/Services/AppUpdate/AppUpdateCoordinator.swift]: (Nit) I would put this directly in the `Version` file`

File: WordPress/Classes/System/WordPressAppDelegate.swift [hunk 2]
```diff
@@ -644,8 +644,8 @@ extension WordPressAppDelegate {
     }
 
     private func checkForAppUpdates() {
-        guard let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String else {
-            DDLogError("No CFBundleShortVersionString found in Info.plist")
+        guard let version = Bundle.main.infoDictionary?["CFBundleVersion"] as? String else {
+            DDLogError("No CFBundleVersion found in Info.plist")
```
Comment (kean) [inline WordPress/Classes/System/WordPressAppDelegate.swift]: (Nit) A good place for `wpAssertionFailure`.

File: WordPress/Classes/Utility/Version.swift [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    /// - Precondition: `prereleaseIdentifiers` can contain only ASCII alpha-numeric characters and "-".
+    /// - Precondition: `buildMetaDataIdentifiers` can contain only ASCII alpha-numeric characters and "-".
+    public init(
+        _ major: Int,
+        _ minor: Int,
+        _ patch: Int,
+        prereleaseIdentifiers: [String] = [],
+        buildMetadataIdentifiers: [String] = []
+    ) {
+        precondition(major >= 0 && minor >= 0 && patch >= 0, "Negative versioning is invalid.")
```
Comment (kean) [inline WordPress/Classes/Utility/Version.swift]: I would suggest replacing `precondition` (crashes in prod) with safe failures and `wpAssertionFailure`.

File: WordPress/Classes/System/WordPressAppDelegate.swift (hunk 2 above)
Comment (momo-ozawa) [reply WordPress/Classes/System/WordPressAppDelegate.swift]: Fixed! [6b62951](https://github.com/wordpress-mobile/WordPress-iOS/pull/23224/commits/6b629513916bf4e7991cb8e30ff0d7e4d6dfe1ba)

File: WordPress/Classes/Services/AppUpdate/AppUpdateCoordinator.swift (hunk 1 above)
Comment (momo-ozawa) [reply WordPress/Classes/Services/AppUpdate/AppUpdateCoordinator.swift]: Fixed! [d97c8d5](https://github.com/wordpress-mobile/WordPress-iOS/pull/23224/commits/d97c8d588727ac648a393b440395585866475158)

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/291">25.0</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/Utility/Version.swift (hunk 3 above)
Comment (momo-ozawa) [reply WordPress/Classes/Utility/Version.swift]: Fixed! [abb2724](https://github.com/wordpress-mobile/WordPress-iOS/pull/23224/commits/abb27246f3ebca40adeaecf46e3b39fbbf5c250a)

**Comment by momo-ozawa** [comment]
@kean Thanks for reviewing - I've addressed all the comments.

**Review by kean — APPROVED** [review]

**Pull request merged**